Design implementation pass of Song Select v2

frenzibyte opened the pull request:
peppynote: took over this PR to knock some sense into things for review. this is a working draft until further notice!

- [x] Depends on https://github.com/ppy/osu/pull/32760
- [x] Depends on https://github.com/ppy/osu/pull/32759
- [x] Depends on https://github.com/ppy/osu/pull/32764

Joehuu committed:
Implement sheared filter text box

Apply minor adjustments to text box implementation

frenzibyte committed:
Define standard font sizes and update new beatmap panel desgins accordingly

Display beatmap card title in styled font

Decrease difficulty author font level in standalone panel

font name changes

frenzibyte committed:
Add shadow support in sheared nub

frenzibyte committed:
Add sheared range slider bar

Fix failing tests

Add star rating variant of range slider bar

frenzibyte committed:
Add sheared collection dropdown

frenzibyte committed:
Add filter control area in song select

peppy commented:
First impressions aren't great:

![osu! 2025-04-08 at 07 18 37](https://github.com/user-attachments/assets/4634a9c4-abeb-4274-ac99-c9abaf3c503e)

![osu! 2025-04-08 at 07 18 22](https://github.com/user-attachments/assets/580ed108-bea3-46b0-8267-c8cd049c7e43)

Left and right halves are overlapping with a pretty normal aspect ratio. @frenzibyte can you confirm this is how you left things and I haven't fucked something up?

peppy commented:
Leaderboard doesn't seem to show placeholders on failures:

https://github.com/user-attachments/assets/45f2b2a8-c04b-4766-8e99-ac8529f6ea07

peppy commented:
Something is going very wrong with this display:

![osu! 2025-04-08 at 07 21 32](https://github.com/user-attachments/assets/7e5d54c1-df37-4eaa-8de2-4a0c4ed78cf9)

frenzibyte commented:
- For the wedge/leaderboard measurements, yes this is how things stayed. I think the fix would be to limit the size of the wedge to not eat the header, roughly 650px last time I tried.
- For leaderboard issues, the design has not been hooked up with a fetching backend yet (see [code](https://github.com/ppy/osu/pull/32715/files#diff-74a160362867682d6b1bc002936cf172bb1b0aed811b956c7a5154990a6012cbR233-R247)), also from previous PR chain:
  
  > This branch in its current state does not hook up the rankings section with actual score fetching backend yet. I've used https://github.com/ppy/osu/pull/32494 temporarily during my own testing and also to record the preview video below. I've left the code necessary to consume the components by the linked PR commented, so that if the PR goes in then attaching it to the components here is simple.
  
  TL;DR: I've made functionality of leaderboard dependant on https://github.com/ppy/osu/pull/32494 

- For the wedge statistics looking broken on 4:3, it is also an issue that I'm not yet sure how to fix and I've left it as a follow-up task / up to discussion (since it's a minor issue surrounded by the gigantic diff of everything design). I'm foreseeing the right-hand difficutly statistics shortening to CS/OD/HP/AR to give breathing room for the left-hand statistics.

peppy committed:
Don't allow focus of difficulty range selector

smoogipoo commented on osu.Game/Screens/SelectV2/WedgeDifficultyDisplay.cs:
```diff
@@ -0,0 +1,379 @@
+                mapperText.Text = beatmap.Value.Metadata.Author.Username;
+            }
+
+            var playableBeatmap = beatmap.Value.GetPlayableBeatmap(ruleset.Value);
```
Without doing a full review, I'll just mention that this is causing a beatmap load on the update thread

peppy committed:
Make `BeatmapFilterControl`'s test actually work

Also

- Fix tests to use a better default display width.
- Avoid adding a second background to test scenes that looks weird.
- Remove seemingly unnecessary padding from `BeatmapFilterControl`.
- Remove weird 1.1 width specification for the control itself.

peppy committed:
Add missing test coverage to `BeatmapInfoWedge`

Also:

- Improve display of difficulty increase/decrease.
- Explain the negative padding and remove arbitrary padding from test.

peppy committed:
General metrics adjust

Global screen padding

peppy committed:
Change sheared dropdowns to use the full available width for items

peppy committed:
Update bars when difficulty adjust is applied to look better

This is mostly just a feeling thing. Will iterate going forward.

peppy committed:
Show placeholders when online playcount/favourites are not available

This avoids causing large layout changes which can distract a user.

peppy committed:
Standardise wedge piece backgrounds and add a bit more texture

peppy committed:
Fix carousel scroll bar extending to underneath toolbar and filter control

frenzibyte committed:
Add wedge background behind filter control

Some margin numbers were further adjusted to ensure the background is
completely covered on the left and top edges

peppy committed:
Change the way `ShearedButton` applies shear

peppy committed:
Add shorter string for "show converts" button

peppy committed:
Fix metrics of filter control

peppy committed:
Second pass of info wedge metrics

peppy committed:
Add redesigned beatmap wedge for new song select

infowedge

Introduce numeric value in beatmap hit count statistics

peppy committed:
Add beatmap details section for new song select

wedgesheader

peppy committed:
Add new song select beatmap rankings section

leaderboard

peppy committed:
Update font specs in line with new naming and structure

peppy committed:
Rename everything

Move remaining v2 tests in non-v2 namespace

frenzibyte committed:
Fix difficulty range slider showing gray on zero

frenzibyte committed:
Update statistics display to work with `BeatmapStatistic`'s new structure

frenzibyte committed:
Refactor metadata display to support truncation properly

peppy committed:
Remove commented things and add a big warning flag that this is probably wrong

I didn't write this code and take no responsibility for this. If there's
any issue, throw it out and rewrite correctly.

peppy committed:
Another pass of code quality things

aaaaaaaaaaaaaaa

peppy committed:
Knock some sense into sheared drawable stuff

peppy committed:
Rename song select v2 classes and namespaces

This aims to bring some conformity to naming to make it easier to
understand component structure for new components.

Renames are pulled out of the song select v2 changes and are more
relevant there due to many new classes being added.

- `V2` suffix is dropped, with v2 components being moved to a relevant V2 namespace.
- Related classes have a prefix of the area they are used.
- Experimenting with using partial/nested classes in the song select v2 implementation.
  Not committing to this yet but want to see how it plays out.
- Moved base carousel components to a generic namespace to avoid confusion with actual beatmap carousel implementation.

peppy committed:
Add keywords for converted setting

peppy committed:
Make some test methods static for future reuse

peppy committed:
Avoid adding arbitrary background in `SongSelectComponentsTestScene`

peppy committed:
Add global screen margin for new screen designs

peppy committed:
Disable masking of `Carousel`

The default for carousels should be unmasked as their usage generally
sees them overflowing outside their main usage area (see `bleed`
variables).

peppy committed:
Add animated fade when online status pill has an unknown status

peppy commented:
I've done what I can with this PR, but I'm still not happy to merge it. There's too many issues with implementation details to even start reviewing. I've voiced concerns with @frenzibyte and he's going to make an attempt to split out the changes here into compartmentalised pieces with the following in mind:

- The "placeholder" functionality from new components should either match the original component in functionality or be removed. One of the major issues I have here is that things were implemented *just enough to make it look like they work* but once you delve into the code you realise that's a facade. A good example is checking [this new code](https://github.com/ppy/osu/blob/321fc45bcc80e2209c0fa866179b37f4ea03f074/osu.Game/Screens/SelectV2/BeatmapTitleWedge_DifficultyStatisticsDisplay.cs#L160-L184) which has functional and code quality issues – the null case doesn't clear existing display, and the check for matching labels creates drawables for no reason.
- Any of the follow up commits and fixes I've made should be squashed down (they don't need to exist). There should only be one (or a handful of relevant) commits per new component
- Each new component should get its own PR, unless multiple components are very closely related and the resultant PR size is <500 LoC changed
- New component PRs should not be touching any existing code (unless it's only a few stray lines)
- All design updates / functionality changes to existing components should be in their own PR(s)

I've already split out some changes in https://github.com/ppy/osu/pull/32817, https://github.com/ppy/osu/pull/32818, https://github.com/ppy/osu/pull/32819. There are around 65 files left to work with.

Will leave this PR open for now, as it may be used to compare on an ongoing basis, but eventually this will be closed with new 100% isolated PRs taking its place.

peppy reviewed (commented):
As commented.

peppy committed:
Use new song select
